python/social_bot/envs/play_ground.py

```diff
@@ -186,6 +200,28 @@ def __init__(self,
         self.reset()
         self.observation_space = self._agent.get_observation_space(
             self._teacher)
+        if self._goal_conditioned:
```

Comment:
Hi Le, 

I think It would be better to put "goal_conditioned" option and get_observation related code to GazeboAgent class in gazebo_agent.py, and keep play_ground.py clean.

Now we are procedding the details of observation and action in class Agent. For example, you can move the observation space building code to the function "GazeboAgent.get_observation_space()".

"play_ground.py" is not supposed to deal with details of ob, action and goal. This should be an option of the GoalTask. 


Add another OrderedDict support in agent.py ( get_observation_space and get_observation ) like:
```

        if self._image_with_internal_states or self._with_language:
            # observation is an OrderedDict with feild ....
            obs = self._create_observation_dict(teacher, sentence_raw)
        elif self._use_image_observation:  # observation is pure image
            obs = self.get_camera_observation()
        elif self._goal_conditioned:
            # observation is OrderedDict with feild ``achieved_goal``, ``desired_goal`` ....
            obs = self._create_gc_observation()
        else:  # observation is pure low-dimentional states
            obs = teacher.get_task_specific_observation(self)
        return obs

```
In self._create_gc_observation(), setup the observation with task_specific_observation, agent position and goal position.
In GoalTask, if self._goal_conditioned, do not include agent position and goal position.

-------

There are mainly 3 class for now: Task, Agent, and Playground:

- task.py contains the details implementation of task
- agent.py defines the interface of agent(like setup joints, build observation and action space, get image or execute action)
- play_ground.py put all of them together.

It might be better to make observation defined by each Task, because only task have all the infomation directly. But to reuse some common codes like building the observation/action space and image (with internal states) case, we defined the observation space in Agent.

Reply:
Good points JT!  Thanks for explaining in detail.   Moved to agent.